src/lib/yang/xpath.lua

```diff
@@ -0,0 +1,260 @@
+      local head
+      head, tail = next_element(tail)
+      if head == nil then return head else return extract_parts(head) end
+   end
+end
+
+-- Finds the grammar node for a fragment in a given grammar.
+local function extract_grammar_node(grammar, fragment)
+   local handlers = {}
+   function handlers.struct () return grammar.members[fragment.name] end
+   function handlers.table ()
+      if grammar.keys[fragment.name] == nil then
+         return grammar.values[fragment.name]
+      else
+         return grammar.keys[fragment.name]
+      end
+   end
+   local errmsg = "Invalid path:"..fragment.name
+   return assert(assert(handlers[grammar.type](), errmsg), errmsg)
+end
+
+-- Converts an XPath path to a lua array consisting of path componants.
+-- A path component can then be resolved on a yang data tree:
+function convert_path(grammar, path)
+   local ret = {}
+   local node = grammar
+   for element in split_path(path) do
+      node = extract_grammar_node(node, element)
+      local luapath = handle(node, element)
+      table.insert(ret, luapath)
+      --node = next_node
+   end
+   return ret
+end
+
+-- Returns a resolver for a paticular schema and *lua* path.
+function resolver(path)
+   local handlers = {}
+   local function handle(data, path)
+      if #path == 0 then return data end
+      local head = table.remove(path, 1)
+      local handler = assert(handlers[head.grammar.type], head.grammar.type)
+      return handler(head, data, path)
+   end
+   function handlers.table(head, data, path)
+      -- The list can either be a ctable or a plain old lua table, if it's the
+      -- ctable then it requires some more work to retrive the data.
+      local data = data[normalize_id(head.name)]
+      if #head.grammar.keys > 1 then
+         error("Multiple key values are not supported!")
+      end
+      if head.grammar.key_ctype then
+         -- It's a ctable and we need to prepare the key so we can lookup the
+         -- pointer to the entry and then convert the values back to lua.
+         local keyname = next(head.keys)
+         local ptype = head.grammar.keys[keyname].argument_type.primitive_type
+         local kparser = valuelib.types[ptype].parse
+         local key = kparser(head.keys[keyname])
+         local ckey = ffi.new(head.grammar.key_ctype)
+         ckey[keyname] = key
+
+         data = data:lookup_ptr(ckey).value
+      else
+         data = data[normalize_id(head.keys[next(head.keys)])]
+      end
+      return handle(data, path)
+   end
+   function handlers.struct(head, data, path)
+      data = data[normalize_id(head.name)]
+      return handle(data, path)
+   end
+   function handlers.array(head, data, path)
+      data = data[normalize_id(head.name)][head.key]
+      return handle(data, path)
+   end
+   function handlers.scalar(head, data, path)
+      if #path ~= 0 then error("Paths can't go beyond leaves.") end
+      return data[normalize_id(head.name)]
+   end
+   return function (data)
```

Comment:
Can this be expressed in such a way that the handler is computed ahead of time?  I.e. do all of the lookups in `handlers` before returning the function.